atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java

```diff
@@ -68,18 +66,98 @@ public SchemaMutationLock(
+        private final Thread heartbeatThread;
+        private final long lockId;
+        private volatile Boolean keepBeating;
+        private volatile long heartbeatCount;
+
+        HeartbeatThread(long lockId) {
+            this.heartbeatThread = new Thread(this::heartbeat);
+            this.lockId = lockId;
+            this.heartbeatCount = 0;
+        }
+
+        void startBeating() {
+            keepBeating = true;
+            heartbeatThread.start();
+        }
+
+        void stopBeating() {
+            keepBeating = false;
+            try {
+                // make sure beating's stopped
+                heartbeatThread.join();
+            } catch (InterruptedException e) {
+                LOGGER.debug("Interrupted while trying to stop the heartbeat heartbeatThread.");
+            }
+        }
+
+        void heartbeat() {
+            while (keepBeating) {
+                try {
+                    clientPool.runWithRetry((FunctionCheckedException<Cassandra.Client, Void, Exception>) client -> {
+                        Column ourUpdate = lockColumnWithStrValue(String.format(
+                                CassandraConstants.GLOBAL_DDL_LOCK_FORMAT, lockId, heartbeatCount + 1));
+
+                        List<Column> expected = ImmutableList.of(lockColumnWithStrValue(String.format(
+                                CassandraConstants.GLOBAL_DDL_LOCK_FORMAT, lockId, heartbeatCount)));
+
+                        CASResult casResult = writeDdlLockWithCas(client, expected, ourUpdate);
+
+                        if (casResult.isSuccess()) {
+                            heartbeatCount++;
+                        } else {
+                            handleForcedLockClear(casResult, lockId, heartbeatCount);
+                        }
+                        return null;
+                    });
+                    Thread.sleep(configManager.getConfig().heartbeatTimePeriodMillis());
+                } catch (Exception e) {
+                    throw Throwables.throwUncheckedException(e);
+                }
+            }
+        }
+    }
 
+    void runWithLock(Action action) {
+        if (!supportsCas) {
+            runWithLockWithoutCas(action);
+            return;
+        }
+
+        long lockId = waitForSchemaMutationLock();
+        heartbeatThread = new HeartbeatThread(lockId);
+        heartbeatThread.startBeating();
         try {
             action.execute();
         } catch (Exception e) {
             throw Throwables.throwUncheckedException(e);
         } finally {
+            heartbeatThread.stopBeating();
             schemaMutationUnlock(lockId);
         }
     }
 
+    private void runWithLockWithoutCas(Action action) {
+        LOGGER.info("Because your version of Cassandra does not support check and set,"
+                + " we will use a java level lock to synchronise schema mutations."
+                + " If this is a clustered service, this could lead to corruption.");
```

Comment:
So technically the only way to know is via the leader block, but even that is not sufficient, because in general I could have a leader block specifying 3 leaders (including myself) and then there could be 2 more clients that specify time and lock blocks that refer to the 3 leaders.  Generally the answer to the question "are other atlas clients running against this underlying keyValueService" is impossible to answer perfectly/completely from the perspective of a single atlas client.